### ISSUE TITLE
no-inferred-empty-object-type not working for function

### ISSUE THREAD

**Issue by OliverJAsh** [palantir/tslint#2947]
### Bug Report

- __TSLint version__: 5.4.3
- __TypeScript version__: 2.4
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
{
    class Wrapper<L> {
        constructor(public val?: L) {}
    }
    let x = new Wrapper(); // expected error, got error
}
{
    const wrapper = <L>(val?: L) => ({ val });
    let x = wrapper(); // expected error, no error
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-inferred-empty-object-type": true
    }
}
```

Potentially related: https://github.com/palantir/tslint/issues/2908

**Comment by rsxdalv** [comment]
I find that to be in line with current design - intersection between undefined and empty object is not equivalent to empty object. What might be of interest though is that this won't generate an error either:
```
{
    const wrapper = <L>(val?: L) => ({ val: val! });
    const x = wrapper(); // Expected error, no error
}
```
Which again, happens due to the inferred type not being empty interface, however, at point of use it becomes one. (Avoidable by disallowing null assertions). Perhaps add an option for catching `{} | undefined`?

**Comment by OliverJAsh** [comment]
Ah, that makes sense. Thanks for the explanation.

> an option for catching {} | undefined?

This would be ne good!

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
